articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,119 +38,389 @@ Result result = session.run(queryString);
+
+It is recommended that this sanitization is done at the client level, before passing it down to the database.
+
+=== Escape Characters
+
+Escape characters invoke an alternative meaning on the following characters in the sequence.
+In Cypher, the definition of string literals, regular expressions and naming of types such as node labels can be done
+with the opening and closing of certain characters that can also be used inside the expression given it is escaped correctly.
+
+In all cases, a backslash is escaped using another backslash.
+
+==== String Literals
+
+String literals are started and ended with either a single quote `'` or a double quote `"`.
+These can be escaped using a backslash `\`.
+
+==== Regular Expressions
+
+Regular expressions are started and ended with a single quote `'` which can be escaped using a backslash.
+
+==== Naming
+
+Node labels, relationship types, parameters and variables can be started and ended with a single backtick ```,
+the backtick is escaped using another backtick ```.
+
+For more information on escape characters, see the Cypher Manual on https://neo4j.com/docs/cypher-manual/current/syntax/expressions/[Expressions]
+and https://neo4j.com/docs/cypher-manual/current/syntax/naming/[Naming rules and recommendations].
+
+=== When Sanitization is Necessary
+
+Node labels, relationship types and parameters may contain non-alphabetic characters, including numbers, symbols and whitespace characters, but must be escaped using backticks.
+For example: `node label with spaces`.
+This means that when dynamically building a query using string concatenation, sanitization needs to be done on the escaping of backticks.
+In Cypher, a backtick is escaped using another backtick `&#96;&#96;`.
+
+Here is an example of a simple dynamic label injection attack:
+
+`query = "MATCH (s:School)-[:IN]->(c:&#96;" + cityName + "&#96;) RETURN s;`
+
+With this query we want to search for all schools that are in a certain city, unfortunately our city names are Node Labels,
+so it isn't possible to parameterize the input.
+
+A possible attack input would be:
+
+`Input = &#96;) RETURN 1 as a UNION MATCH (n) RETURN 1 WITH DISTINCT true AS haxxored MATCH (n) DETACH DELETE n; \\`
+
+The backtick escapes the label name context and the parentheses closes the node.
+The `UNION` here then ensures that a match is made, as if the first `MATCH` statement doesn't return anything, the next part of the query won't be run.
+The `WITH` reduces the result set down to one row and then the final part will delete everything in the database.
+
+This attack was not possible to avoid using parameterization.
+To avoid this attack sanitization must be used.
+It is possible to add validation to the query as well, in this case, validating the city name is a real city name before passing it into the database, and rejecting it otherwise.
```

Comment:
```suggestion
It is possible to add validation to the user input (?) as well, in this case, validating the city name is a real city name before passing it into the database, and rejecting it otherwise.
```